src/knot_cloud.c

```diff
@@ -120,127 +96,74 @@ static int map_routing_key_to_msg_type(const char *routing_key)
+		has_err = parser_is_key_str_or_null(json_str,
+			KNOT_JSON_FIELD_ERROR) ? false : true;
 
-		msg->list = parser_request_to_list(jso);
-		if (!msg->list) {
-			l_error("Malformed JSON message");
-			goto err;
-		}
+	if (has_err) {
+		l_error("Ill-formed JSON message");
+		knot_cloud_msg_destroy(msg);
+		return NULL;
+	}
 
+	switch (msg->type) {
+	case UPDATE_MSG:
+		msg->list = parser_update_to_list(json_str);
+		has_err = msg->list ? false : true;
+		break;
+	case REQUEST_MSG:
+		msg->list = parser_request_to_list(json_str);
+		has_err = msg->list ? false : true;
 		break;
 	case REGISTER_MSG:
-		msg->device_id = parser_get_key_str_from_json_obj(jso,
-			KNOT_JSON_FIELD_DEVICE_ID);
-		if (!msg->device_id ||
-				!parser_is_key_str_or_null(jso,
-						KNOT_JSON_FIELD_ERROR)) {
-			l_error("Malformed JSON message");
-			goto err;
-		}
-
-		msg->token = parser_get_key_str_from_json_obj(jso,
+		msg->token = parser_get_key_str_from_json_str(json_str,
 			KNOT_JSON_FIELD_DEVICE_TOKEN);
-		if (!msg->token) {
-			l_error("Malformed JSON message");
-			goto err;
-		}
-
-		msg->error = parser_get_key_str_from_json_obj(jso,
-			KNOT_JSON_FIELD_ERROR);
+		has_err = msg->token ? false : true;
 		break;
 	case UNREGISTER_MSG:
-		msg->device_id = parser_get_key_str_from_json_obj(jso,
-			KNOT_JSON_FIELD_DEVICE_ID);
-		if (!msg->device_id ||
-				!parser_is_key_str_or_null(jso,
-					KNOT_JSON_FIELD_ERROR)) {
-			l_error("Malformed JSON message");
-			goto err;
-		}
-
-		msg->error = parser_get_key_str_from_json_obj(jso,
-			KNOT_JSON_FIELD_ERROR);
-		break;
 	case AUTH_MSG:
-		msg->device_id = parser_get_key_str_from_json_obj(jso,
-			KNOT_JSON_FIELD_DEVICE_ID);
-		if (!msg->device_id ||
-				!parser_is_key_str_or_null(jso,
-					KNOT_JSON_FIELD_ERROR)) {
-			l_error("Malformed JSON message");
-			goto err;
-		}
-
-		msg->error = parser_get_key_str_from_json_obj(jso,
-			KNOT_JSON_FIELD_ERROR);
-		break;
 	case SCHEMA_MSG:
-		msg->device_id = parser_get_key_str_from_json_obj(jso,
-			KNOT_JSON_FIELD_DEVICE_ID);
-		if (!msg->device_id ||
-				!parser_is_key_str_or_null(jso,
-					KNOT_JSON_FIELD_ERROR)) {
-			l_error("Malformed JSON message");
-			goto err;
-		}
-
-		msg->error = parser_get_key_str_from_json_obj(jso,
-			KNOT_JSON_FIELD_ERROR);
 		break;
 	case LIST_MSG:
-		msg->device_id = NULL;
-		msg->list = parser_queue_from_json_array(jso,
-					knot_cloud_device_array_foreach);
-		if (!msg->list || !parser_is_key_str_or_null(jso,
-				KNOT_JSON_FIELD_ERROR)) {
-			l_error("Malformed JSON message");
-			goto err;
-		}
-
-		msg->error = parser_get_key_str_from_json_obj(jso,
-			KNOT_JSON_FIELD_ERROR);
+		msg->list = parser_queue_from_json_array(json_str,
+				create_device_item);
+		has_err = msg->list ? false : true;
 		break;
 	case MSG_TYPES_LENGTH:
 	default:
 		l_error("Unknown event %s", routing_key);
-		goto err;
+		has_err = true;
+		break;
+	}
+
+	if (has_err) {
+		l_error("Ill-formed JSON message");
+		knot_cloud_msg_destroy(msg);
```

Comment:
Free the allocated memory for the string on `msg->token` inside this function, since the `parser_get_key_str_from_json_str` function is allocating a string.

Reply:
in order to do this a had to change token (const char * to char *)